0009-S-sdk-resiliency.md

```diff
@@ -0,0 +1,122 @@
+  - Timeout and retry needs to be implemented at the user's application.
+2. Add retry only
+  - Undetermined behavior when sidecar is down or too slow. For example, the Java SDK simply gets stuck forever if there is no response from the sidecar (tested with ToxiProxy).
+3. Let each SDK decide how to handle this.
+  - Inconsistent behavior and configuration for resiliency, requiring system admins to know specifics of each SDK.
+
+* Are there any trade-offs being made? (space for time, for example)
+1. Simplification of retry policy, having an opinionated setting for most configuration.
+2. No support for Circuit Breaking or API health check prior to calling the Dapr API.
+
+* What advantages / disadvantages does this proposal have? 
+Pros:
+- Bring consistency and simple set of configuration points that work cross SDKs
+- Document expected behavior for SDKs regarding timeout and retries
+
+Cons:
+- See trade-offs mentiond above.
+
+## Implementation Details
+
+### Design
+
+* `DAPR_API_MAX_RETRIES` defines the maximum number of retries, SDKs can determine which strategy will be implemented (linear, exponential backoff, etc). `0` is the default value and means no retry (same for negative values).
+* `DAPR_API_TIMEOUT_SECONDS` defines the maximum waiting time to connect and receive a response for an HTTP or gRPC call. Defaults to `0`. `0` (or negative) are handled as "undefined" and calls might hang forever on the client side. This setting is the timeout for each API invocation and not the timeout of the aggregated time for retries. This setting can be used without retries.
+* All environment variables can be overwritten via parameters to the Dapr client or at a per-request basis, in the following order (higher priority on top):
+  1. Per-request parameter
+  2. Parameter when instantiating a Dapr client object
+  3. Properties or any other language specific configuration framework.
+  4. Environment variables
+* SDK to retry if error is on connection.
+* SDK to retry in case of the following retriable codes:
+  * gRPC: DEADLINE_EXCEEDED, UNAVAILABLE.
+  * HTTP: 408, 429 (respect `Retry-After` header), 500, 502, 503, 504
+* The same client should still be usable if the API goes down but is restored after any arbitrary amount of time. In other words, the unavailability of the Dapr API should not require the application to restart.
+
+#### Example of implementation
+
+https://github.com/dapr/java-sdk/pull/889
+
+### Feature lifecycle outline
+
+* Compatability guarantees
+Retries and timeouts should be disabled by default.
+
+* Deprecation / co-existence with existing functionality
+If customers prefer to have a more fine tuned resiliency logic, they can still achieve so by disabling the SDK resiliency and use a 3rd party library to handle retries with custom logic.
+
+* Feature flags
+Retries and timeouts are disabled by default with the value `0`.
```

Comment:
Is it also worth calling out the infinite retry case here? Do we want to support it? Will it be a `-1` like runtime? Will it be a special policy?

Reply:
I am not adding infinite retries as a thing now. It does not mean it cannot be added later. Thoughts?

Reply:
Let me set it as -1 to infinity in this proposal to avoid a breaking change in the future. Default is still 0 for SDKs.